doc: Updates contributing file with code and test best practices (#1666)
mongodb/terraform-provider-mongodbatlas: CONTRIBUTING.md

Reviewer: I think I commented this before but this file is too large and a split should be considered

Author: I think you did, ticket created: https://jira.mongodb.org/browse/INTMDB-1394

- Lastly, in the terminal run `terraform init` again and this time terraform will pull provider version from `tf_cache` network mirror. You can confirm this by noting the `Terraform has been successfully initialized! Using mongodb/mongodbatlas Vx.x.x from the shared cache directory` message.  

> ## Code and Test Best Practices

Reviewer 2: Some follow up idea:

Most of the upstream repositories address that by defining "folder structure" chapter where each significant folder is documented and visualized in the tree. 

Additionally you can reference go/terraform standards so readers can quickly gain context based on their understanding of the ecosystem

Author: thanks! i'll add it to the JIRA ticket above.

> - Each resource (and associated data sources) is in a package in `internal/service`.

Reviewer: needs clarification on what "service" means, is it going to be the openapi tags?

Author: no, i was the only one for it, team decision was one per resource as stated here

> - There can be multiple helper classes and they can also be used from other resources, e.g. `common_advanced_cluster.go` defines functions that are also used from other resources using `advancedcluster.FunctionName`.

Reviewer 3: ```suggestion
- There can be multiple helper files and they can also be used from other resources, e.g. `common_advanced_cluster.go` defines functions that are also used from other resources using `advancedcluster.FunctionName`.
```

> - Unit and Acceptances tests are in the same `_test.go` file. They are not in the same package as the code tests, e.g. `advancedcluster` tests are in `advancedcluster_test` package so coupling is minimized.

Reviewer 2: Format for brevity consider adding chapters:

```suggestion

## Code Changes
...

## Unit and Acceptance tests
```

Contributing is usually focused on guiding new person to do specific tasks. 

Instead of

> - Unit and Acceptances tests are in the same `_test.go` file. They are not in the same package as the code tests, e.g. `advancedcluster` tests are in `advancedcluster_test` package so coupling is minimized.

Consider:

> Create *_test.go file for both Unit and Acceptance tests.
For example see: link.

Consider reader empathy:

> package so coupling is minimized.

do we need that info for the new contributions?

Reviewer 2: Those aren't blocking PR. Just some nits and CONTRIBUTING.md format tips for future.

- Migration tests are in `_migration_test.go` files.

> - Helper methods can have their own tests, e.g. `common_advanced_cluster_test.go` has tests for `common_advanced_cluster.go`.

Reviewer 4: ```suggestion
- Helper methods must have their own tests, e.g. `common_advanced_cluster_test.go` has tests for `common_advanced_cluster.go`.
```

- `internal/testutils/acc` contains helper test methods for Acceptance and Migration tests.

> - Tests that need the provider binary like End-to-End tests don’t belong to the source code packages and go in a different structure.

Reviewer 4: ```suggestion
- Tests that need the provider binary like End-to-End tests don’t belong to the source code packages and go in `test/e2e`.
```
